### PR TITLE
Add support for Fragments

### PR DESCRIPTION
**Pull request by wangdaliu**
https://trello.com/c/z4KMdFUm/16-4-add-support-for-fragments

### CODE REVIEW

File: affirm/src/main/java/com/affirm/android/AffirmFragment.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+
+        webView = view.findViewById(R.id.webview);
+        progressIndicator = view.findViewById(R.id.progressIndicator);
+
+        if (getActivity() != null) {
+            AffirmUtils.debuggableWebView(getActivity());
```
Comment (kevindavmoore) [inline affirm/src/main/java/com/affirm/android/AffirmFragment.java:80]: Should this be only for debug mode?

File: README.md [hunk 2]
```diff
@@ -262,6 +262,30 @@ Affirm.trackOrderConfirmed(MainActivity.this, trackModel());
...
+    // vcn checkout
+    // In your activity, you need implements Affirm.VcnCheckoutCallbacks
+    Affirm.startCheckout(this, R.id.container, checkoutModel(), true);
+
+    // promotion
+    AffirmPromotionButton affirmPromotionButton = findViewById(R.id.promo);
+    Affirm.configureWithAmount(R.id.container, affirmPromotionButton, PromoPageType.PRODUCT, PRICE, true);
+
+    // site modal
+    // In your activity, you need implements Affirm.PrequalCallbacks
```
[the remark this answers is not in the document]
Comment (wangdaliu) [reply README.md]: Fixed

File: README.md [hunk 3]
```diff
@@ -262,6 +262,30 @@ Affirm.trackOrderConfirmed(MainActivity.this, trackModel());
...
+    Affirm.showSiteModal(FragmentUsagesActivity.this, R.id.container, "5LNMQ33SEUYHLNUC");
+
+    // product modal
+    // In your activity, you need implements Affirm.PrequalCallbacks
```
[the remark this answers is not in the document]
Comment (wangdaliu) [reply README.md]: Fixed

File: affirm/src/main/java/com/affirm/android/AffirmFragment.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentManager;
+
+public abstract class AffirmFragment extends Fragment implements AffirmWebChromeClient.Callbacks {
+
+    static final String TAG_PREFIX = "AffirmFragment";
+
+    WebView webView;
```
[the remark this answers is not in the document]
Comment (wangdaliu) [reply affirm/src/main/java/com/affirm/android/AffirmFragment.java]: Fixed

File: affirm/src/main/java/com/affirm/android/Affirm.java [hunk 5]
```diff
@@ -614,6 +730,68 @@ public static void configureWithAmount(@NonNull final AffirmPromotionButton prom
...
+     */
+    public static void configureWithAmount(
+            @IdRes int containerViewId,
+            @NonNull final AffirmPromotionButton promotionButton,
+            @Nullable final String promoId,
+            @Nullable final PromoPageType pageType,
+            final BigDecimal amount,
+            final boolean showCta) {
+
+        final View.OnClickListener onClickListener = v -> {
```
[the remark this answers is not in the document]
Comment (wangdaliu) [reply affirm/src/main/java/com/affirm/android/Affirm.java]: Fixed

File: affirm/src/main/java/com/affirm/android/AffirmFragment.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    static void addFragment(FragmentManager fragmentManager, @IdRes int containerViewId,
+                            @NonNull Fragment fragment, @NonNull String tag) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            try {
+                fragmentManager.beginTransaction().add(containerViewId, fragment, tag).commitNow();
+            } catch (IllegalStateException | NullPointerException e) {
+                fragmentManager.beginTransaction().add(containerViewId, fragment, tag).commit();
+                try {
+                    fragmentManager.executePendingTransactions();
```
[the remark this answers is not in the document]
Comment (wangdaliu) [reply affirm/src/main/java/com/affirm/android/AffirmFragment.java]: Fixed

File: affirm/src/main/java/com/affirm/android/AffirmFragment.java (hunk 1 above)
Comment (wangdaliu) [reply affirm/src/main/java/com/affirm/android/AffirmFragment.java]: In the `debuggableWebView` method, I have checked the debug mode

File: affirm/src/main/java/com/affirm/android/ModalFragment.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                HTTPS_PROTOCOL + AffirmPlugins.get().baseUrl(),
+                html, TEXT_HTML, UTF_8, null);
+    }
+
+    private String initialHtml() {
+        String html;
+        try {
+            final InputStream ins = getResources().openRawResource(type.templateRes);
+            html = AffirmUtils.readInputStream(ins);
+        } catch (IOException e) {
```
Comment (kevindavmoore) [inline affirm/src/main/java/com/affirm/android/ModalFragment.java:142]: Shouldn't you close this stream?

File: affirm/src/main/java/com/affirm/android/PrequalFragment.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    @Override
+    public void onDetach() {
+        listener = null;
+        super.onDetach();
+    }
+
+    @Override
+    public void onWebViewError(@NonNull ConnectionException error) {
+        AffirmTracker.track(PREQUAL_WEBVIEW_FAIL, ERROR, null);
```
Comment (kevindavmoore) [inline affirm/src/main/java/com/affirm/android/PrequalFragment.java:106]: Add error?

File: affirm/src/main/java/com/affirm/android/VcnCheckoutFragment.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    boolean useVCN() {
+        return true;
+    }
+
+    @Override
+    InnerCheckoutCallback innerCheckoutCallback() {
+        return new InnerCheckoutCallback() {
+            @Override
+            public void onError(@NonNull AffirmException exception) {
+                AffirmTracker.track(VCN_CHECKOUT_CREATION_FAIL, ERROR, null);
```
Comment (kevindavmoore) [inline affirm/src/main/java/com/affirm/android/VcnCheckoutFragment.java:99]: Exception?

File: affirm/src/main/java/com/affirm/android/VcnCheckoutFragment.java [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+                        TEXT_HTML, UTF_8, null);
+            }
+        };
+    }
+
+    private String initialHtml(@NonNull CheckoutResponse response) {
+        String html;
+        try {
+            final InputStream ins = getResources().openRawResource(R.raw.affirm_vcn_checkout);
+            html = AffirmUtils.readInputStream(ins);
```
Comment (kevindavmoore) [inline affirm/src/main/java/com/affirm/android/VcnCheckoutFragment.java:121]: Close stream?

File: samples-java/src/main/java/com/affirm/samples/FragmentUsagesActivity.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+        findViewById(R.id.vcnCheckout).setOnClickListener(v -> {
+            try {
+                Affirm.startCheckout(this, R.id.container, checkoutModel(), true);
+            } catch (Exception e) {
+                Toast.makeText(getBaseContext(), "VCN Checkout failed, reason: " + e.toString(), Toast.LENGTH_SHORT).show();
+            }
+        });
+
+        findViewById(R.id.siteModalButton).setOnClickListener(v -> Affirm.showSiteModal(FragmentUsagesActivity.this, R.id.container, "5LNMQ33SEUYHLNUC"));
```
Comment (kevindavmoore) [inline samples-java/src/main/java/com/affirm/samples/FragmentUsagesActivity.java:55]: I assume this is a dummy number?

**Comment by wangdaliu** [comment]
There are too many conflicts. Close this and create a new PR.

**Pull request closed without merging**